### DISCUSSION TITLE
Can a sequencing primitive like futures;:join! be created that automatically links a chain of operations?

### DISCUSSION THREAD

**Discussion by farnoy** [tokio-rs/tokio-uring#79] [Ideas]
To avoid round trips to the runtime and user code between operations that can be submitted ahead of time, can we automate the creation of chains?

It would make it possible to convert subsequent operations that do not have data dependencies, but are relying on precise exceptions, into a chain.

Before
```rs
let (res, buf1) = file.write_at(buf1, first_offset).await;
res?;
let (res, buf2) = file.write_at(buf2, second_offset).await;
res?;
```

After
```rs
let ((res1, buf1), (res2, buf2)) =
    tokio_uring::chain!(file.write_at(buf1, first_offset), file.write_at(buf2, second_offset));

res1?;
// if res1 was an error, res2 is also guaranteed to be -ECANCELED
// if we used `futures::join!`, res2 could have succeeded despite the error in res1
res2?; 
```

This should enable fewer round trips to the runtime while preserving precise exceptions and linear execution.

**Comment by Noah-Kennedy** [comment]
Funnily enough I've been thinking of this as well, moreso related to event linkage.

**Reply by Noah-Kennedy** [reply]
This would require a lot of API changes, but that's fine, this is pretty alpha level stuff, so breakage should be expected.

**Reply by farnoy** [reply]
Since you mentioned API changes, this would work best if we could reuse the same buffer for multiple operations in the chain. My write example wouldn't benefit from it, but if you wanted to write and then read in a single chain, you'd want to reuse the same buffer. This is currently impossible because the values are being moved.

I don't want to derail the discussion, but there's also the whole buffer registration aspect. It would be awesome if futures could loan buffers from the runtime (and it would automatically allocate & register a new one if none are available) and automatically returning them on Drop.

**Reply by Noah-Kennedy** [reply]
These are good points. The current API was designed to try and be close to the original tokio API, but I think it doesn't really do a good job of letting users take advantage of the more powerful, low level capabilities of tokio-uring.
